### DISCUSSION TITLE
Enable Langfuse can expose metrics in Prometheus format

### DISCUSSION THREAD

**Discussion by gyliu513** [langfuse/langfuse#2508] [Ideas]
### Describe the feature or potential improvement

Now there are some customers want to build their own dashboard, and they want all of the metrics should be exposed in prometheus format, so that they can build their own dashboard by scraping the prometheus format metrics and build their own dashboard.

There are already some open source projects exposing their metrics in prometheus format, include the following:
- https://github.com/opencost/opencost?tab=readme-ov-file#opencost--your-favorite-open-source-cost-monitoring-tool-for-kubernetes-and-cloud-spend
- https://github.com/NVIDIA/dcgm-exporter?tab=readme-ov-file#quickstart
- https://sustainable-computing.io/design/architecture/#kepler-exporter

Enabling this can help enlarge the ecosystem for langfuse to integrate with some other observability platforms, like IBM Instana, Dynatrace, Datadog, New Relic etc.

Thanks for providing such a great place to enable such discussions!

**Comment by marcklingen** [comment]
thanks for adding this here Guangya 🙏

**Reply by paulnegz** [reply]
I'll love to contribute the community by implementing this change! Looking forward to feedback, open to make changes - https://github.com/langfuse/langfuse/pull/8157/files @marcklingen @gyliu513

**Reply by Steffen911** [reply]
@paulnegz Can you add details on why the current flow seems insufficient? And what options were considered to implement this?

**Comment by Steffen911** [comment]
@gyliu513 Could you add more details on which metrics you're interested in?
The suggestion of @paulnegz focuses on metrics about the Langfuse instance, less on business metrics like the observation cost, etc. I think depending on the needs we'll have to consider different implementation approaches here.

**Comment by gyliu513** [comment]
@Steffen911 I do not have specified metrics, but what I cared is having a Prometheus endpoint can expose metrics, so that I can get metrics from this endpoint. Prometheus endpoint is the key to integrate with some 3rd party observability platforms. Thanks!

**Reply by Steffen911** [reply]
@gyliu513 But I think its usefulness heavily depends on the kinds of metrics you'd want to see (and also the effort and reasonable-ness to implement one).

Adding Langfuse specific metrics like `events processed`, `queue length`, etc. should be easy and I see how that's reasonable for self-hosted infrastructure monitoring.

We've also heard from users that want to see metrics about the entities being tracked within Langfuse, e.g. LLM costs tagged by user etc. This would require lots of customization and would likely be expensive to implement and maintain. So depending on the actual needs the feasibility of this request is very different.

**Reply by gyliu513** [reply]
@Steffen911 I got you, for AI Observability, we should follow otel semantic convention at https://github.com/open-telemetry/semantic-conventions/blob/main/model/gen-ai/registry.yaml to get those metrics.

**Comment by skalebag** [comment]
@Steffen911 I have been trying to use langfuse to get the metrics in Grafana via prometheus, I did reach out to the langfuse team and even though we cannot export the metrics as of now they mentioned this is a future feature that might get added, I wanted to particularly get metrics around the application behavior like tokens cost system metrics errors latency etc, If this feature get added or we get an sdk it would be amazing, Open to help and contribute.

**Reply by Steffen911** [reply]
@skalebag I believe this will be hard to implement generically in a performant way. In the meantime, you could build this yourself using our metrics-api: https://langfuse.com/docs/metrics/features/metrics-api.

**Comment by 0xfaizan** [comment]
+1 , would be very helpful if Langfuse could provide an endpoint for LLM observability in /metrics format.
